text/3-entrees/storage-api/vec-set.md

```diff
@@ -0,0 +1,145 @@
+not already a member and the membership limit has not been reached. We check for these two
+conditions first, and then insert the new member only after we are sure it is safe to do so.
+
+```rust, ignore
+pub fn add_member(origin) -> DispatchResult {
+	let new_member = ensure_signed(origin)?;
+
+	let mut members = Members::<T>::get();
+	ensure!(members.len() < MAX_MEMBERS, Error::<T>::MembershipLimitReached);
+
+	// We don't want to add duplicate members, so we check whether the potential new
+	// member is already present in the list. Because the list is always ordered, we can
+	// leverage the binary search which makes this check O(log n).
+	match members.binary_search(&new_member) {
+		// If the search succeeds, the caller is already a member, so just return
+		Ok(_) => Err(Error::<T>::AlreadyMember.into()),
+		// If the search fails, the caller is not a member and we learned the index where
+		// they should be inserted
+		Err(index) => {
+			members.insert(index, new_member.clone());
+			Members::<T>::put(members);
+			Self::deposit_event(RawEvent::MemberAdded(new_member));
+			Ok(())
+		}
+	}
+}
+```
+
+If it turns out that the caller is not already a member, the binary search will fail. In this case
+it still returns the index into the `Vec` at which the member would have been stored had he been
+present. We then use this information to insert the member at the appropriate location, thus
+maintaining a sorted `Vec`.
+
+## Removing a Member
+
+Removing a member is entirely straightforward. We begin by looking for the caller in the list. If
+the caller is not present, there is no work to be done. If the caller is present, the search
+algorithm returns her index, and she can be removed.
+
+```rust, ignore
+fn remove_member(origin) -> DispatchResult {
+	let old_member = ensure_signed(origin)?;
+
+	let mut members = Members::<T>::get();
+
+	// We have to find out where, in the sorted vec the member is, if anywhere.
+	match members.binary_search(&old_member) {
+		// If the search succeeds, the caller is a member, so remove her
+		Ok(index) => {
+			members.remove(index);
+			Members::<T>::put(members);
+			Self::deposit_event(RawEvent::MemberRemoved(old_member));
+			Ok(())
+		},
+		// If the search fails, the caller is not a member, so just return
+		Err(_) => Err(Error::<T>::NotMember.into()),
+	}
+}
+```
+
+## Performance
+
+Now that we have built our set, let's analyze its performance in some common operations.
+
+### Membership Check
+
+In order to check for the presence of an item in a `vec-set`, we make a single storage read, decode
+the entire vector, and perform a binary search.
+
+DB Reads: O(1) Decoding: O(n) Search: O(log n)
+
+### Updating
+
+Updates to the set, such as adding and removing members as we demonstrated, requires first
+performing a membership check. It also requires re-encoding the entire `Vec` and storing it back
+in the database.
+
+DB Writes: O(1) Encoding: O(n)
```

Comment:
Mutating a `Vec` in memory can re-allocate, which in turn means all the members are copied over to the new allocation. So adding a member is `O(1)` until the `Vec` is at capacity, and then it's `O(n)` for that addition, and then back to `O(1)` again. It's complicated! :)

Reply:
Good call. I added a bit about that.